scheduler: internal data structure optimization

Huang-Wei opened the pull request:
**What type of PR is this?**

/kind feature

**What this PR does / why we need it**:

Optimize the data structure of ~~`topologySpreadMap`~~`podSpreadCache` to reduce memory and runtime overhead, and also boost the performance of predicates and preemption of EvenPodsSpread.

Here is the new structure:

```go
// podSpreadCache combines tpKeyToCriticalPaths and tpPairToMatchNum
// to represent:
// (1) critical paths where the least pods are matched on each spread constraint.
// (2) number of pods matched on each spread constraint.
type podSpreadCache struct {
	// We don't record all critical paths here.
	// If there is only one criticalPath, criticalPaths[0] always holds the minimum matchNum;
	// If there are multiple criticalPaths, keep arbitrary 2 of them holding the minimum matchNum.
	tpKeyToCriticalPaths map[string][2]*criticalPath
	// tpPairToMatchNum is keyed with topologyPair, and valued with the number of matching pods.
	tpPairToMatchNum map[topologyPair]int32
}
```

**CAVEAT**: this PR (in particular the `[2]critialPath` structure) can work is based on the implementation of the current preemption algorithm:
Fact 1: we only preempt pods on the **same** node, instead of pods on multiple nodes.
Fact 2: each node is evaluated on a separate copy of the metadata during its preemption cycle.
If we plan to turn to a more complex algorithm like "arbitrary pods on multiple nodes", this enhancement needs to be revisited.

**Which issue(s) this PR fixes**:

Part of #77284. Resolve [comment1](https://github.com/kubernetes/kubernetes/pull/77760#issuecomment-513126887) and [comment2](https://github.com/kubernetes/kubernetes/pull/79062#discussion_r305470724).

**Special notes for your reviewer**:

The performance looks good (~6X in time, ~4X in space). Let's focus on functionality reviewing.

#### Before

```
single-constraint-zone-8        100   16717361 ns/op  12650462 B/op  129741 allocs/op
single-constraint-node-8        100   17020951 ns/op  12674237 B/op  132412 allocs/op
two-constraints-zone-node-8      50   22534697 ns/op  16043751 B/op  203792 allocs/op

single-constraint-zone-8        100	  17250549 ns/op  12646159 B/op  129737 allocs/op
single-constraint-node-8        100	  14650029 ns/op  12672823 B/op  132416 allocs/op
two-constraints-zone-node-8      50	  21257976 ns/op  16043452 B/op  203793 allocs/op

single-constraint-zone-8        100	  14535822 ns/op  12651489 B/op  129740 allocs/op
single-constraint-node-8        100	  14415079 ns/op  12672163 B/op  132414 allocs/op
two-constraints-zone-node-8     100	  18504999 ns/op  16041346 B/op  203793 allocs/op
```

#### After

```
single-constraint-zone-8        500    2523942 ns/op  2796400 B/op   60027 allocs/op
single-constraint-node-8        500    2863317 ns/op  2984588 B/op   60051 allocs/op
two-constraints-zone-node-8     300	   5202418 ns/op  5753595 B/op  120058 allocs/op

single-constraint-zone-8        500	   2574097 ns/op  2795724 B/op   60027 allocs/op
single-constraint-node-8        500	   2824376 ns/op  2985523 B/op   60051 allocs/op
two-constraints-zone-node-8     300	   4965061 ns/op  5751401 B/op  120059 allocs/op

single-constraint-zone-8        500	   2488182 ns/op  2796486 B/op   60027 allocs/op
single-constraint-node-8        500	   2922101 ns/op  2988127 B/op   60051 allocs/op
two-constraints-zone-node-8     300	   5169896 ns/op  5751460 B/op  120059 allocs/op
```


**Does this PR introduce a user-facing change?**:

```release-note
NONE
```

/sig scheduling
/assign @ahg-g @alculquicondor 
cc/ @bsalamat @liu-cong @draveness @krmayankk

Huang-Wei committed:
Add Benchmark test for EvenPodsSpread predicate

- go test k8s.io/kubernetes/pkg/scheduler/algorithm/predicates -benchmem -run=^$ -bench .

Huang-Wei commented:
/hold
There is a data race.

alculquicondor commented on pkg/scheduler/algorithm/predicates/metadata.go:
```diff
@@ -393,98 +395,105 @@ func (m *topologyPairsPodSpreadMap) addPod(addedPod, preemptorPod *v1.Pod, node
+		//    is definitely greater than "paths[1].matchNum".
+		if v == paths[0].topologyValue {
+			paths[0].matchNum = curMatch
+			// paths[0].matchNum has increased by 1,
```
in this case, at least 2 keys had the same count, say (4, 4). After adding, we get (5, 4), so we need to swap to (4, 5). This works if those were the 2 only keys with that count. If we had (4, 4, 4), then after adding, the state should be (4, 4, 5).

Huang-Wei replied:
`[2]critialPath` in preemption case doesn't always hold the **minimum** two paths. We only need to guarantee `path[0]` always holds the minimum matchNum, and that's also our predicate algorithm relies on - compare to the minMatch.

So in your case, it's fine that at some point it becomes [4,5] (one of the original 4 gets lost). At anytime we must keep the minMatch tracked well.

alculquicondor reviewed (commented):
I stopped the review because I think this implementation doesn't work. Specifically, it's not enough to rely in 2 critical paths when we can have "almost" arbitrary removal/addition of pods (see the comments for details).

We have 2 options:

- Implement a priority queue (maybe we can reuse some already implemented),
- Do a linear search for the minimum each time we add a node. Maybe this is enough, because preemption is not a common case. But we can support this with performance tests.

Additionally, some memory can be saved if we store a map from node to topology keys (instead of pod to topologyKeys).

Huang-Wei commented on pkg/scheduler/algorithm/predicates/metadata.go:
```diff
@@ -67,17 +67,24 @@ type topologyPairsMaps struct {
-// (2) how existing pods match incoming pod on its spread constraints.
+// (1) critical paths where the least pods are matched on each spread constraint.
+// (2) number of pods matched on each spread constraint.
 type topologyPairsPodSpreadMap struct {
```
Or `podSpreadCache`?

alculquicondor replied:
sgtm

Huang-Wei commented on pkg/scheduler/algorithm/predicates/metadata.go:
```diff
@@ -279,18 +272,36 @@ func getTPMapMatchingSpreadConstraints(pod *v1.Pod, nodeInfoMap map[string]*sche
+		key := constraints[i].TopologyKey
+		m.tpKeyToCriticalPaths[key] = [2]*criticalPath{{matchNum: math.MaxInt32}, {matchNum: math.MaxInt32}}
+	}
+	// TODO(Huang-Wei): it may be optimized to O(log(n)) as it's a topK (K=2) problem.
```
You're right, it can't be faster than O(n). I will remove the comment.

Huang-Wei commented:
TestPreemptWithPermitPlugin flake +1: #81238

Huang-Wei commented:
/retest

Huang-Wei commented:
Recent (3rd) commit basically involves a number of changes due to the renaming, but it doesn't change the core logic/algorithm of this PR.

Huang-Wei commented:
/retest

alculquicondor commented on pkg/scheduler/algorithm/predicates/metadata.go:
```diff
@@ -279,18 +272,35 @@ func getTPMapMatchingSpreadConstraints(pod *v1.Pod, nodeInfoMap map[string]*sche
+}
+
+// sortCriticalPaths is only served for testing purpose.
+func (c *podSpreadCache) sortCriticalPaths() {
```
can we move this to `metadata_test.go`?

alculquicondor reviewed (commented):
As discussed, let's leave the 2 critical paths implementation, documenting its caveats (in PR description and code comments).

alculquicondor commented on pkg/scheduler/algorithm/predicates/metadata.go:
```diff
@@ -67,17 +67,24 @@ type topologyPairsMaps struct {
+	// We don't record all critical paths here.
+	// If there is only one criticalPath, criticalPaths[0] always holds the minimum matchNum;
+	// If there are multiple criticalPaths, keep arbitrary 2 of them holding the minimum matchNum.
+	tpKeyToCriticalPaths map[string][2]*criticalPath
```
Also, we are choosing int32 here, whereas we are using int64 for the priorities

Huang-Wei replied:
To make them consistent, I'd like to change int64 in priorities to int32, any preference?

alculquicondor replied:
Since this is counting pods, that should be fine.

Huang-Wei commented on pkg/scheduler/algorithm/predicates/metadata.go:
```diff
@@ -337,21 +347,16 @@ func newTopologyPairsMaps() *topologyPairsMaps {
 func (m *topologyPairsMaps) addTopologyPair(pair topologyPair, pod *v1.Pod) {
 	podFullName := schedutil.GetPodFullName(pod)
-	m.addTopologyPairWithoutPods(pair)
+	if m.topologyPairToPods[pair] == nil {
```
Correct :)

Huang-Wei commented:
/retest

Huang-Wei commented:
/retest

alculquicondor commented on pkg/scheduler/algorithm/predicates/metadata.go:
```diff
@@ -473,15 +440,15 @@ func (c *podSpreadCache) clone() *podSpreadCache {
 	}
 	for tpKey, paths := range c.tpKeyToCriticalPaths {
-		copyPaths := [2]*criticalPath{
+		copyPaths := criticalPaths{
```
nit: `criticalPaths{paths[0], paths[1]}`

Huang-Wei replied:
Neat! It's an Array so we can simply put its value (as a copy) here.

alculquicondor commented on pkg/scheduler/algorithm/predicates/metadata_test.go:
```diff
@@ -1635,7 +1635,7 @@ func TestPodSpreadCache_removePod(t *testing.T) {
 				deletedPod = tt.deletedPod
 			}
-			podSpreadCache.removePod(deletedPod, tt.preemptor, tt.nodes[tt.nodeIdx])
+			podSpreadCache.updatePod(deletedPod, tt.preemptor, tt.nodes[tt.nodeIdx], -1)
```
not a big fan of the numbers. Maybe just add `addPod` and `removePod` that call `updatePod`? But this is a nit.

Huang-Wei replied:
Updated.

alculquicondor commented:
/lgtm
unless you want to implement the nits

Huang-Wei commented:
/hold cancel

Address updated and commits (except for the benchmark one) have been squashed.

@alculquicondor PTAL.

Huang-Wei commented:
/hold

@ahg-g wants to do a quick review.

alculquicondor reviewed (commented):
/lgtm

Huang-Wei commented:
/retest

Huang-Wei committed:
Optimize internal data structure of EvenPodsSpread

- Rename 'topologyPairsPodSpreadMap' to 'podSpreadCache'
- New struct `criticalPaths criticalPaths`
- Add unified method `*criticalPaths.update()` for:
    - regular update
    - addPod in preemption case
    - remotePod in preemption case

ahg-g commented on pkg/scheduler/algorithm/predicates/metadata.go:
```diff
@@ -67,17 +67,67 @@ type topologyPairsMaps struct {
+// CAVEAT: the reason that `[2]criticalPath` can work is based on the implementation of current
+// preemption algorithm - we only preempt pods on the same node, instead of pods on multiple nodes.
+// If we plan to turn to a more complex algorithm like "arbitrary pods on multiple nodes", this
+// structure needs to be revisited.
```
we need to also clarify that this works also because each node is evaluated on a separate copy of the metadata.

Huang-Wei replied:
SG.

ahg-g reviewed (commented):
Thanks! this is clearly more efficient and easier to reason about compared to the original implementation.

Huang-Wei commented on pkg/scheduler/algorithm/predicates/metadata.go:
```diff
@@ -67,17 +67,67 @@ type topologyPairsMaps struct {
+type podSpreadCache struct {
+	// We don't record all critical paths here.
+	// If there is only one criticalPath, criticalPaths[0] always holds the minimum matchNum;
+	// If there are multiple criticalPaths, keep arbitrary 2 of them holding the minimum matchNum.
```
That sounds too short... How about:

```
	// We record 2 critical paths instead of all critical paths here.
	// criticalPaths[0].matchNum always holds the minimum matching number.
	// criticalPaths[1].matchNum is always greater or equal to criticalPaths[0].matchNum,
	// but it's not guaranteed to be the 2nd minimum match number.
```

Huang-Wei commented on pkg/scheduler/algorithm/predicates/metadata.go:
```diff
@@ -67,17 +67,67 @@ type topologyPairsMaps struct {
+	} else {
+		// `tpVal` doesn't exist
+		if num < paths[0].matchNum {
+			// update paths[1] with paths[1]
```
Done.

Huang-Wei commented on pkg/scheduler/algorithm/predicates/metadata.go:
```diff
@@ -67,17 +67,67 @@ type topologyPairsMaps struct {
+		j := (i + 1) % 2 // the other index
+		// adjust i and j to make sure paths[0] always holds the minimum match num.
+		if (i < j && paths[i].matchNum > paths[j].matchNum) ||
+			(i > j && paths[i].matchNum < paths[j].matchNum) {
```
It looks good.

Another (fancy but may hard to read) way I though was: `i < j == paths[i].matchNum > paths[j].matchNum`.

alculquicondor replied:
It looks much better now :)

Huang-Wei commented:
@ahg-g Comments addressed. PTAL.

Huang-Wei commented:
/retest

ahg-g commented:
/lgtm
/approve

Huang-Wei commented:
@ahg-g @alculquicondor Commits squashed. Please help to relabel /lgtm.

Thanks for the efforts reviewing this!

Huang-Wei commented:
/hold cancel

ahg-g commented:
/lgtm

Huang-Wei commented:
/retest

Huang-Wei commented:
/retest
